rfd/0106-machine-id-anonymous-telemetry.md

```diff
@@ -0,0 +1,157 @@
+---
+authors: Noah Stride (noah.stride@goteleport.com)
+state: draft
+---
+
+# RFD 0106 - Machine ID Anonymous Agent Telemetry
+
+## Required Approvers
+
+* Engineering: @zmb3
+* Product: @klizhentas && @xinding33
+
+## What
+
+Collect anonymous usage information about invocations of the Machine ID agent.
+
+Out of scope is the collection of usage information from the Auth server.
+
+## Why
+
+It is currently difficult to determine the adoption of Machine ID, and for what
+use-cases they use Machine ID for. This makes product decisions about 
+determining areas of focus more difficult.
+
+Basic anonymous telemetry from the `tbot` agent will provide helpful
+information on adoption without compromising privacy.
+
+## Details
+
+### Event collection
+
+For now, a single event on startup will be submitted by `tbot`.
+
+This will be implemented as part of `tool/tbot` rather than `lib/tbot`, meaning
+that events will be submitted when using the `tbot` binary but not when
+embedding the `tbot` library within another binary. This means data will not
+be polluted by internal uses of the `tbot` library (e.g the operator).
+
+Event collection and submission will be started concurrently to the `tbot`
+functionality, and should not impede the primary function of `tbot`. In the
+case of failure, a warning message should be omitted. The routine will read in
+the configuration of `tbot`, extract the relevant values and encode these within
+an event protobuf.
+
+### Event submission
+
+Events will be submitted directly to an unauthenticated `tbot` specific endpoint
+of the `prehog` service:
+
+```protobuf
+syntax = "proto3";
+
+package prehog.v1alpha;
+
+import "google/protobuf/timestamp.proto";
+
+message SubmitTbotEventRequest {
+  // UUID identifying that tbot session. This is future-proofing for if we
+  // decide to add multiple events in future, and need to tie them together.
+  string distinct_id = 1;
+  // optional, will default to the ingest time if unset
+  google.protobuf.Timestamp timestamp = 2;
+
+  oneof event {
+    // See the events section for the fields included within startup.
+    TbotStartEvent start = 3;
+  }
+}
+
+message SubmitTbotEventResponse {}
+
+service TbotReportingService {
+  rpc SubmitTbotEvent(SubmitTbotEventRequest) returns (SubmitTbotEventResponse) {}
+}
+```
+
+As only a single event will be submitted, batching is not of concern.
+
+### Event storage
+
+Events received by `prehog` will be encoded in a PostHog compatible format
+and then submitted to PostHog.
+
+Tbot events will be share the same project as clusters and website events.
+
+### Consent
+
+Users will explicitly opt-in to anonymous usage telemetry with an environment
+variable, e.g:
+
+```sh
+# Send anonymous telemetry on startup about the usage of tbot in order
+# to help us understand what parts of Machine ID to improve.
+# Find out more at goteleport.com/awesome-docs-page
+export TBOT_ANONYMOUS_TELEMETRY=1
```

Comment:
```suggestion
export TELEPORT_ANONYMOUS_TELEMETRY=1
```

Since with the default GHA (for example), you don't even really need to know what a `tbot` is.